File: backend/payment_crypto.py
"""AES-256-GCM encryption for payment provider configuration secrets.

Purpose-built for the Nomba (and any future payment provider) admin
config, kept separate from the existing MFA Fernet system
(backend/server.py's _cloud_encrypt/_cloud_decrypt) deliberately:

  - That system derives its key from JWT_SECRET (a shared secret used
    for session tokens too) with no key-versioning. Payment provider
    credentials are a higher-stakes secret category and deserve their
    own dedicated master key, so rotating JWT_SECRET never silently
    breaks (or is coupled to) decrypting stored payment credentials,
    and vice versa.
  - AES-256-GCM specifically (not AES-128-CBC+HMAC, which is what
    Fernet uses) per the explicit spec for this migration.
  - Key-version support, so the master key can be rotated later without
    a flag day that breaks every previously-encrypted row.

Uses `cryptography` (already a project dependency, cryptography==46.0.7
per backend/requirements.txt) — a proven library, not hand-rolled crypto.

Master key source: PAYMENT_CONFIG_ENCRYPTION_KEY environment variable
only. Never stored in the database, never editable from the admin
dashboard, never logged. Must decode (after base64) to exactly 32 bytes
(AES-256). Generate one with:
    python3 -c "import secrets,base64; print(base64.b64encode(secrets.token_bytes(32)).decode())"
"""
from __future__ import annotations

import base64
import os
import secrets as _secrets
from typing import Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

# v-prefixed key version, so a future re-key can add "v2": "<base64 key>"
# alongside this one and decrypt() can dispatch on the stored version tag
# without invalidating existing rows.
_CURRENT_KEY_VERSION = "v1"


class PaymentCryptoNotConfigured(RuntimeError):
    """Raised when PAYMENT_CONFIG_ENCRYPTION_KEY is missing or malformed.
    Callers must fail closed (refuse to save/read payment credentials),
    never silently fall back to plaintext."""


def _load_master_key() -> bytes:
    raw = os.environ.get("PAYMENT_CONFIG_ENCRYPTION_KEY", "").strip()
    if not raw:
        raise PaymentCryptoNotConfigured(
            "PAYMENT_CONFIG_ENCRYPTION_KEY is not set. Payment provider "
            "credentials cannot be encrypted or decrypted without it. Generate "
            "one with: python3 -c \"import secrets,base64; "
            "print(base64.b64encode(secrets.token_bytes(32)).decode())\" "
            "and set it in the server environment (never in the database, "
            "never in the admin dashboard)."
        )
    try:
        key = base64.b64decode(raw, validate=True)
    except Exception as exc:
        raise PaymentCryptoNotConfigured(
            f"PAYMENT_CONFIG_ENCRYPTION_KEY is not valid base64: {exc}"
        ) from exc
    if len(key) != 32:
        raise PaymentCryptoNotConfigured(
            f"PAYMENT_CONFIG_ENCRYPTION_KEY must decode to exactly 32 bytes "
            f"(AES-256), got {len(key)} bytes."
        )
    return key


def encrypt_secret(plaintext: str) -> dict:
    """Encrypts a single secret value. Returns a dict safe to store as a
    MongoDB subdocument: {"v": key_version, "n": base64 nonce, "ct": base64
    ciphertext+tag}. AESGCM's ciphertext output already includes the auth
    tag appended -- no separate tag field needed."""
    if plaintext is None:
        plaintext = ""
    key = _load_master_key()
    aesgcm = AESGCM(key)
    nonce = _secrets.token_bytes(12)  # 96-bit nonce, random per encryption
    ct = aesgcm.encrypt(nonce, plaintext.encode("utf-8"), None)
    return {
        "v": _CURRENT_KEY_VERSION,
        "n": base64.b64encode(nonce).decode("ascii"),
        "ct": base64.b64encode(ct).decode("ascii"),
    }


def decrypt_secret(enc: Optional[dict]) -> str:
    """Decrypts a dict produced by encrypt_secret(). Returns "" for
    None/empty input (never-configured case) rather than raising, so
    callers can treat "not configured" and "decrypt failed" distinctly:
    this returns "" for the former, raises for the latter (tampered/
    wrong-key ciphertext) rather than silently returning "" for both --
    a payment integration must never treat "credentials corrupted"
    the same as "credentials not set"."""
    if not enc:
        return ""
    key = _load_master_key()
    version = enc.get("v")
    if version != _CURRENT_KEY_VERSION:
        raise PaymentCryptoNotConfigured(
            f"Encrypted payment secret has unknown key version {version!r} "
            f"(expected {_CURRENT_KEY_VERSION!r}). A key-rotation migration "
            "is needed before this value can be decrypted."
        )
    try:
        nonce = base64.b64decode(enc["n"])
        ct = base64.b64decode(enc["ct"])
        aesgcm = AESGCM(key)
        pt = aesgcm.decrypt(nonce, ct, None)
        return pt.decode("utf-8")
    except Exception as exc:
        raise RuntimeError(
            "Failed to decrypt a stored payment secret -- either "
            "PAYMENT_CONFIG_ENCRYPTION_KEY has changed since it was "
            "encrypted, or the stored value is corrupted/tampered. Refusing "
            "to silently treat this as 'not configured'."
        ) from exc


def mask_preview(plaintext: str) -> str:
    """Same preview convention already used for paystack_key_preview /
    onesignal_api_key_preview in the existing admin settings response --
    never return the real secret to the browser."""
    if not plaintext:
        return ""
    if len(plaintext) <= 12:
        return "set"
    return f"{plaintext[:6]}...{plaintext[-4:]}"


def is_configured() -> bool:
    """Cheap check for whether PAYMENT_CONFIG_ENCRYPTION_KEY is present
    and well-formed, without needing an actual secret to decrypt --
    used by /admin routes and startup checks to report a clear
    remediation message instead of an opaque failure."""
    try:
        _load_master_key()
        return True
    except PaymentCryptoNotConfigured:
        return False

File: backend_node/src/services/paymentCrypto.ts
import { createCipheriv, createDecipheriv, randomBytes } from "node:crypto";
import { env } from "../env.js";

/**
 * Port of backend/payment_crypto.py -- AES-256-GCM encryption for payment
 * provider configuration secrets. Separate master key from the MFA Fernet
 * system (payment_crypto.ts derives from PAYMENT_CONFIG_ENCRYPTION_KEY
 * only, never JWT_SECRET) and separate cipher (AES-256-GCM, not AES-128-CBC).
 */

const CURRENT_KEY_VERSION = "v1";

export class PaymentCryptoNotConfigured extends Error {}

export interface EncryptedSecret {
  v: string;
  n: string;
  ct: string;
}

function loadMasterKey(): Buffer {
  const raw = env.PAYMENT_CONFIG_ENCRYPTION_KEY.trim();
  if (!raw) {
    throw new PaymentCryptoNotConfigured(
      "PAYMENT_CONFIG_ENCRYPTION_KEY is not set. Payment provider credentials cannot be encrypted or decrypted without it.",
    );
  }
  let key: Buffer;
  try {
    key = Buffer.from(raw, "base64");
  } catch (exc) {
    throw new PaymentCryptoNotConfigured(`PAYMENT_CONFIG_ENCRYPTION_KEY is not valid base64: ${String(exc)}`);
  }
  if (key.length !== 32) {
    throw new PaymentCryptoNotConfigured(`PAYMENT_CONFIG_ENCRYPTION_KEY must decode to exactly 32 bytes (AES-256), got ${key.length} bytes.`);
  }
  return key;
}

/** Port of payment_crypto.py:73 `encrypt_secret`. AESGCM's ciphertext already includes the auth tag appended -- matches `cryptography`'s AESGCM.encrypt() output format. */
export function encryptSecret(plaintext: string | null): EncryptedSecret {
  const text = plaintext ?? "";
  const key = loadMasterKey();
  const nonce = randomBytes(12);
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  const ciphertext = Buffer.concat([cipher.update(text, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return {
    v: CURRENT_KEY_VERSION,
    n: nonce.toString("base64"),
    ct: Buffer.concat([ciphertext, tag]).toString("base64"),
  };
}

/** Port of payment_crypto.py:91 `decrypt_secret`. Returns "" for null/empty input; throws for tampered/wrong-key ciphertext (never conflated). */
export function decryptSecret(enc: EncryptedSecret | null | undefined): string {
  if (!enc) return "";
  const key = loadMasterKey();
  if (enc.v !== CURRENT_KEY_VERSION) {
    throw new PaymentCryptoNotConfigured(`Encrypted payment secret has unknown key version ${JSON.stringify(enc.v)} (expected "${CURRENT_KEY_VERSION}").`);
  }
  try {
    const nonce = Buffer.from(enc.n, "base64");
    const raw = Buffer.from(enc.ct, "base64");
    const ciphertext = raw.subarray(0, raw.length - 16);
    const tag = raw.subarray(raw.length - 16);
    const decipher = createDecipheriv("aes-256-gcm", key, nonce);
    decipher.setAuthTag(tag);
    const plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
    return plaintext.toString("utf8");
  } catch (exc) {
    throw new Error(
      "Failed to decrypt a stored payment secret -- either PAYMENT_CONFIG_ENCRYPTION_KEY has changed since it was encrypted, or the stored value is corrupted/tampered.",
      { cause: exc },
    );
  }
}

/** Port of payment_crypto.py:124 `mask_preview`. */
export function maskPreview(plaintext: string | null | undefined): string {
  if (!plaintext) return "";
  if (plaintext.length <= 12) return "set";
  return `${plaintext.slice(0, 6)}...${plaintext.slice(-4)}`;
}

/** Port of payment_crypto.py:135 `is_configured`. */
export function isConfigured(): boolean {
  try {
    loadMasterKey();
    return true;
  } catch (e) {
    if (e instanceof PaymentCryptoNotConfigured) return false;
    throw e;
  }
}
